src/Microsoft.IdentityModel.Protocols.OpenIdConnect/OpenIdConnectProtocolValidator.cs

```diff
@@ -178,110 +185,222 @@ public TimeSpan NonceLifetime
 
             if (idToken.Payload.Iss == null)
-                LogHelper.Throw(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10309, JwtRegisteredClaimNames.Iss.ToLowerInvariant(), idToken), typeof(OpenIdConnectProtocolException), EventLevel.Error);
+                LogHelper.Throw(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10314, JwtRegisteredClaimNames.Iss.ToLowerInvariant(), idToken), typeof(OpenIdConnectProtocolException), EventLevel.Error);
 
             // sub is optional by default
             if (RequireSub && (string.IsNullOrWhiteSpace(idToken.Payload.Sub)))
-                LogHelper.Throw(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10309, JwtRegisteredClaimNames.Sub.ToLowerInvariant(), idToken), typeof(OpenIdConnectProtocolException), EventLevel.Error);
+                LogHelper.Throw(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10314, JwtRegisteredClaimNames.Sub.ToLowerInvariant(), idToken), typeof(OpenIdConnectProtocolException), EventLevel.Error);
 
             // optional claims
             if (RequireAcr && string.IsNullOrWhiteSpace(idToken.Payload.Acr))
-                LogHelper.Throw(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10312, idToken), typeof(OpenIdConnectProtocolException), EventLevel.Error);
+                LogHelper.Throw(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10315, idToken), typeof(OpenIdConnectProtocolException), EventLevel.Error);
 
             if (RequireAmr && string.IsNullOrWhiteSpace(idToken.Payload.Amr))
-                LogHelper.Throw(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10313, idToken), typeof(OpenIdConnectProtocolException), EventLevel.Error);
+                LogHelper.Throw(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10316, idToken), typeof(OpenIdConnectProtocolException), EventLevel.Error);
 
             if (RequireAuthTime && string.IsNullOrWhiteSpace(idToken.Payload.AuthTime))
-                LogHelper.Throw(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10314, idToken), typeof(OpenIdConnectProtocolException), EventLevel.Error);
+                LogHelper.Throw(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10317, idToken), typeof(OpenIdConnectProtocolException), EventLevel.Error);
 
             if (RequireAzp && string.IsNullOrWhiteSpace(idToken.Payload.Azp))
-                LogHelper.Throw(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10315, idToken), typeof(OpenIdConnectProtocolException), EventLevel.Error);
+                LogHelper.Throw(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10318, idToken), typeof(OpenIdConnectProtocolException), EventLevel.Error);
+
+            // if user sets the custom validator, we call the delegate. The default checks for multiple audiences and azp are not executed.
+            if (this.IdTokenValidator != null)
```

Comment:
We may want to consider placing this above all checks, thereby giving users complete control using a delegate. If users want complete control, they will have to use the derive model.